compiler/codegen/CCData.hpp

```diff
@@ -0,0 +1,207 @@
+       * @param[In] value The value to put.
+       * @param[In] key Optional. The key to map the resulting index to. Without a key the index is the only reference to the data.
+       * @param[Out] index The index that refers to the value.
+       */
+      template <typename T>
+      bool put(const T value, const key_t * const key, index_t &index);
+
+      /**
+       * @brief Puts the given value in the table, optionally mapped to the given key (if any), aligned to the given boundary (in bytes), and returns the index to the value. Synchronized.
+
+       * @param[In] value Optional. A pointer to the value to put. If null, no data will be copied but the space will be allocated none the less.
+       * @param[In] sizeBytes The size of the value pointed to.
+       * @param[In] alignmentBytes The alignment (in bytes) to align the value to.
+       * @param[In] key Optional. The key to map the resulting index to. Without a key the index is the only reference to the data. If the key is already mapped to an index the operation will return the index and true, but no data will be written.
+       * @param[Out] index The index that refers to the value.
+       * @return True if the value was placed in the table, or the key was already mapped to an index, false otherwise.
+       */
+      bool put(const uint8_t * const value, const size_t sizeBytes, const size_t alignmentBytes, const key_t * const key, index_t &index);
+
+      /**
+       * @brief Gets the value referred to by the index from the table.
+
+       * @param[In] T The type of the value to get from the table. The type must be TriviallyCopyable. A static assertion checks for the std::is_trivially_copyable<T> type trait on capable compilers.
+       * @param[In] index The index that refers to the value.
+       * @param[out] value A reference to the value to write the result to.
+       * @return True if the value was placed in the table, false otherwise.
+       */
+      template <typename T>
+      bool get(const index_t index, T &value) const;
+
+      /**
+       * @brief Gets the value referred to by the index from the table.
+
+       * @param[In] index The index that refers to the value.
+       * @param[in] value A pointer to the value to write the result to. This parameter is ignored unless this function returns true.
+       * @param[In] sizeBytes The size (in bytes) of the value pointed to. This parameter is ignored unless this function returns true.
+       * @return True if the index refers to an existing value, false otherwise.
+       */
+      bool get(const index_t index, uint8_t * const value, const size_t sizeBytes) const;
+
+      /**
+       * @brief Gets a pointer to the value referred to by the index from the table.
+
+       * @param[In] T The type of the value to get from the table. The type need not be TriviallyCopyable, since this function doesn't do any copying, but it probably should be for symmetry with the put() functions.
```

Comment:
Does the sentence about trivially copyable types actually add any value?

Reply:
No, upon further consideration there's no need for it to be TriviallyCopyable because there's no need to use `put()` to place the value in the first place. You can reserve the space, get a pointer, and construct in place to deal with non-TriviallyCopyable types, so the guidance is a bit misleading. I'll remove it.